### PR TITLE
[MGDOBR-536] replace ingress with knative kafka broker

### PR DESCRIPTION
**Pull request by r00ta**
[MGDOBR-536](https://issues.redhat.com/browse/MGDOBR-536)

This PR aims to replace the ingress application with knative

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/BridgeIngressServiceImpl.java [hunk 1]
```diff
@@ -174,4 +115,93 @@ public Secret fetchBridgeIngressSecret(BridgeIngress bridgeIngress) {
...
+
+        return existing;
+    }
+
+    @Override
+    public KnativeBroker fetchOrCreateBridgeIngressBroker(BridgeIngress bridgeIngress, ConfigMap configMap) {
+        KnativeBroker expected = templateProvider.loadBridgeIngressBrokerTemplate(bridgeIngress);
+        expected.getSpec().getConfig().setName(configMap.getMetadata().getName());
+        expected.getSpec().getConfig().setNamespace(configMap.getMetadata().getNamespace());
+        expected.getMetadata().getAnnotations().replace("x-kafka.eventing.knative.dev/external.topic", configMap.getData().get("topic.name"));
```
[the remark this answers is not in the document]
Comment (r00ta) [reply shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/BridgeIngressServiceImpl.java]: sure, it's still wip

**Comment by r00ta** [comment]
We have to configure the dead letter queue in the broker due to https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/788

**Comment by matzew** [comment]
> We have to configure the dead letter queue in the broker due to https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/788

Our strong recommendation here would be to use a `KafkaSink` as the dead letter sink for the broker

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceImpl.java [hunk 2]
```diff
@@ -154,12 +95,15 @@ public void deleteBridgeIngress(BridgeDTO bridgeDTO) {
...
+        Secret expected = templateProvider.loadBridgeIngressSecretTemplate(bridgeIngress, TemplateImportConfig.withDefaults());
+        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_BOOTSTRAP_SERVERS_SECRET,
+                Base64.getEncoder().encodeToString(bridgeDTO.getKafkaConnection().getBootstrapServers().getBytes()));
+        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_USER_SECRET, Base64.getEncoder().encodeToString(bridgeDTO.getKafkaConnection().getClientId().getBytes()));
+        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_PASSWORD_SECRET, Base64.getEncoder().encodeToString(bridgeDTO.getKafkaConnection().getClientSecret().getBytes()));
+        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_PROTOCOL_SECRET, Base64.getEncoder().encodeToString(bridgeDTO.getKafkaConnection().getSecurityProtocol().getBytes()));
+        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_TOPIC_NAME_SECRET, Base64.getEncoder().encodeToString(bridgeDTO.getKafkaConnection().getTopic().getBytes()));
+        // TODO: refactor with data from DTO
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceImpl.java:105]: To be fixed

File: .github/workflows/Operator-CI.yaml [hunk 3]
```diff
@@ -63,20 +63,23 @@ jobs:
...
       - name: Run integration tests
+        env:
+          IT_SHARD_KAFKA_BOOTSTRAP_SERVERS: ${{ secrets.IT_SHARD_KAFKA_BOOTSTRAP_SERVERS }}
+          IT_SHARD_KAFKA_USER: ${{ secrets.IT_SHARD_KAFKA_USER }}
+          IT_SHARD_KAFKA_PASSWORD: ${{ secrets.IT_SHARD_KAFKA_PASSWORD }}
+          IT_SHARD_KAFKA_TOPIC_NAME: ${{ secrets.IT_SHARD_KAFKA_TOPIC_NAME }}
         uses: ./.github/actions/maven
         with:
-          maven-command: clean verify -Pcucumber -Dparallel
+          maven-command: clean verify -Pcucumber -Dparallel -Dit.shard.kafka.bootstrap.servers=$IT_SHARD_KAFKA_BOOTSTRAP_SERVERS -Dit.shard.kafka.user=$IT_SHARD_KAFKA_USER -Dit.shard.kafka.password=$IT_SHARD_KAFKA_PASSWORD -Dit.shard.kafka.topic.name=$IT_SHARD_KAFKA_TOPIC_NAME -Dit.shard.kafka.protocol=SASL_SSL -Dit.shard.kafka.sasl.mechanism=PLAIN
```
Comment (danielezonca) [inline .github/workflows/Operator-CI.yaml:82]: Is this because now the Kafka Broker directly connect to a cluster while before we were not doing this connection?

File: kustomize/base-openshift/shard/cluster-role.yaml [hunk 4]
```diff
@@ -1,44 +1,45 @@
...
+      - "*"
   - apiGroups:
       - apiextensions.k8s.io
     resources:
       - customresourcedefinitions
     verbs:
-      - get
-      - list
+      - "get"
+      - "list"
```
Comment (danielezonca) [inline kustomize/base-openshift/shard/cluster-role.yaml:18]: Why these changes?

File: kustomize/base/shard/resources/deploy.yaml [hunk 5]
```diff
@@ -9,10 +9,10 @@ spec:
   replicas: 1
   template:
     spec:
-      serviceAccountName: shard-operator
+      serviceAccountName: shard-operator-knative
       containers:
         - image: event-bridge-shard-operator
-          imagePullPolicy: IfNotPresent
+          imagePullPolicy: Always
```
Comment (danielezonca) [inline kustomize/base/shard/resources/deploy.yaml:15]: Why? I think we should stay with `IfNotPresent`

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java [hunk 6]
```diff
@@ -91,44 +88,26 @@ public UpdateControl<BridgeIngress> reconcile(BridgeIngress bridgeIngress, Conte
...
-        // Create Service
-        Service service = bridgeIngressService.fetchOrCreateBridgeIngressService(bridgeIngress, deployment);
-        if (service.getStatus() == null) {
-            LOGGER.debug("Ingress service BridgeIngress: '{}' in namespace '{}' is NOT ready", bridgeIngress.getMetadata().getName(),
+        if (path == null) {
+            LOGGER.info("Knative broker resource BridgeIngress: '{}' in namespace '{}' is NOT ready", bridgeIngress.getMetadata().getName(),
                     bridgeIngress.getMetadata().getNamespace());
             bridgeIngress.getStatus().markConditionFalse(ConditionType.Ready);
-            bridgeIngress.getStatus().markConditionTrue(ConditionType.Augmentation, ConditionReason.ServiceNotReady);
+            bridgeIngress.getStatus().markConditionTrue(ConditionType.Augmentation, ConditionReason.DeploymentNotAvailable); // TODO: replace with KnativeBrokerNotReady
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java:102]: To be fixed

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java [hunk 7]
```diff
@@ -163,26 +126,38 @@ public UpdateControl<BridgeIngress> reconcile(BridgeIngress bridgeIngress, Conte
...
+        // Linked resources are automatically deleted except for Authorization Policy and the ingress due to https://github.com/istio/istio/issues/37221
+
+        // Knative broker needs the dependent secret in order to be deleted properly https://coreos.slack.com/archives/CEXRYS5QC/p1649752951251439?thread_ts=1649752173.045369&cid=CEXRYS5QC
+        kubernetesClient.resources(KnativeBroker.class)
+                .inNamespace(bridgeIngress.getMetadata().getNamespace())
+                .withName(bridgeIngress.getMetadata().getName())
+                .delete();
+
+        // Since the authorizationPolicy has to be in the istio-system namespace due to https://github.com/istio/istio/issues/37221
+        // we can not set the owner reference. We have to delete the resource manually.
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java:146]: Can you please create a ticket to track this? I would like to review it in the future

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/KubernetesNetworkingService.java [hunk 8]
```diff
@@ -46,13 +52,25 @@ public EventSource buildInformerEventSource(String component) {
...
+                .inNamespace(service.getMetadata().getNamespace())
+                .withName(bridgeIngress.getMetadata().getName())
+                .get();
 
         if (existing == null || !expected.getSpec().equals(existing.getSpec())) {
-            client.network().v1().ingresses().inNamespace(service.getMetadata().getNamespace()).withName(service.getMetadata().getName()).createOrReplace(expected);
+            try {
+                LOGGER.info(new ObjectMapper().writeValueAsString(expected));
+            } catch (JsonProcessingException e) {
+                e.printStackTrace();
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/networking/KubernetesNetworkingService.java:68]: Replace with logging/re-throwing the exception

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/knative/KnativeCondition.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+ * <p/>
+ * <em>
+ * Objects may report multiple conditions, and new types of conditions may be added in the future or by 3rd party controllers. Therefore,
+ * conditions are represented using a list/slice of objects, where each condition has a similar structure.
+ * This collection should be treated as a map with a key of type.
+ * </em>
+ *
+ * @see <a href="https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties">Kubernetes API Conventions - Typical Status
+ *      Properties</a>
+ */
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/knative/KnativeCondition.java:34]: Is this class copied from somewhere? I don't think you wrote this javadoc :)

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java [hunk 10]
```diff
@@ -163,26 +126,38 @@ public UpdateControl<BridgeIngress> reconcile(BridgeIngress bridgeIngress, Conte
...
     }
 
     @Override
     public DeleteControl cleanup(BridgeIngress bridgeIngress, Context context) {
         LOGGER.info("Deleted BridgeIngress: '{}' in namespace '{}'", bridgeIngress.getMetadata().getName(), bridgeIngress.getMetadata().getNamespace());
 
-        // Linked resources are automatically deleted
+        // Linked resources are automatically deleted except for Authorization Policy and the ingress due to https://github.com/istio/istio/issues/37221
+
+        // Knative broker needs the dependent secret in order to be deleted properly https://coreos.slack.com/archives/CEXRYS5QC/p1649752951251439?thread_ts=1649752173.045369&cid=CEXRYS5QC
```
Comment (danielezonca) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeIngressController.java:139]: Can you please create a ticket to track this? It should be solved in the future

File: kustomize/overlays/dev/secrets/event-bridge-manager-secrets.yaml [hunk 11]
```diff
@@ -6,7 +6,7 @@ metadata:
...
+    EVENT_BRIDGE_KAFKA_BOOTSTRAP_SERVERS: AgA4f35assjSnt8aUT5ONa3TZfzIxoTGMN7ja41nB9SGap7j+eWus2LUWIMQaNdvaD7p7tYW8EpTPEBCDAN2MaMz4Dahky1TrhtmpKPCMZUOjiAdi5PsT5h31a14wM6H2Ad8J5EZFZ7CPLC/9WTo6N4KhrUrrT0rzV7ufLwvax1U/7YyySv+lsWmSYASFcSP+9vu1FC/wzoKiUzWlGpCMZmF50zlU5XJrmKy5jmmFTI9CwCRKyzQN0fTB2c2FcKVYd/meMBSOxXi9lURDcSt8XnPLLbDxVayfNGq+o9Giqm3uEo6NmtccgLiU14Xo8tmRa/W+3WkOttbQ91/CpIlrH7hqMgEUPSw8vHLiT7WDPOKTuZAWqiy4EHlW5rrpSp3cY08IbnBnivwBmeux3iTipyAbrnSTuv73/e9ihOVQnIsD3hYJIkJkVj8/afe+hidzH19W/SxDw/9Nba8tI8DVTYHSCdpHIFmzvq/gSqk6FdL/F2s3iw7o6S4eu5f2n5fSbuLmRnMj0NVskkruwuufzVttIIv2zagpMZ5I8SZ4KzvRiE3cXJLmIPW+VRo8aqjds62HFzUpwef/N9FgWxcFwOYT6+6HqZ4OuBNUfrrzzFuiPpezzPDLaC9zkQzkzdbzFvNaSFXK37LYPj4xRLUuBv57Uf49xnB/eHbQ4JZbT1klv2F9+a0z1TqjYpiy80jkWzb0sZCTq91tirOWIjcBoH/c9fUNkqyrXrosSmtmmj4xwdrKsveUI+jSwPFXFcSvj8oC1Omf1o579M=
```
[the remark this answers is not in the document]
Comment (r00ta) [reply kustomize/overlays/dev/secrets/event-bridge-manager-secrets.yaml:9]: because at the moment `EVENT_BRIDGE_KAFKA_BOOTSTRAP_SERVERS` starts with `https` but knative does not accept it, as kafka does not use an `https` protocol

File: kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml [hunk 12]
```diff
@@ -9,7 +9,7 @@ spec:
...
+    EVENT_BRIDGE_KAFKA_BOOTSTRAP_SERVERS: AgAsFFT+KBw/oVHimSoeIPw0xPg7m/8zQVi6HqsUre4ebpklnURAUBnJB2PKCA0ef/+i/Ot9xljQC65KyjYhvrzR+xIZHYlzbnYXVxL4Zyu9FG9aOxc/Km6ngBlsMVns/mwJa4l1JCWPsstX8RiuVlSaXdcOjJ4HNU2kWXOISOkHg4hXfkZUUKoNHtd0QZx002iEr81kiGKvomFhpvjpHAWQujypPN2VksPt9e0UkPbIGHHOndJocjWb80q47agv252P0DgKuEJxGW0yZ09ZPLVWDNShBUupVl1WXR52+78eALUfXg8hfhDYCOErcn9I9n1oH0H+cT+djQpetVKBzMJB0Qo06U4msW5v0WenWITRErbXKZeqrF1erZO628WFjT4fO5sOVkPNFu162CQo6U4aeDAVp7jevyOXOeMs9gYZLIC7My8O/oEWPVvH2Q30GZdowWc67kalPhm9WvP5CgQuZSVzq1Bqv2E+fe/5A1xHZE3t6q/mmbj08EjYQKqQ/991atDB/nE+KqrBApl//nfuw4ceLwtlMOKEqzsEGA0PGJh3Jry1asbO2Q9OXz6OnUMkrluCxO0g1IbtXercB2C9enGw5mBZRSVWk0wiRWwDcOYgM8hijHoDdqqH0RBeSY3B3kvLn0qgdakEcWtDYJ2/uEWg7KfBZbGWx7l7uMUb1w85CIB1r/ASC8Z/+hAUkWr/vAlrFQSSN+6n5aE3/9jj6WKFnvQaWkFbbglUTxMBR7TA5KhJOcWW6DSfxFYyzYZi86CCuayl84A=
```
[the remark this answers is not in the document]
Comment (r00ta) [reply kustomize/overlays/stable/secrets/event-bridge-manager-secrets.yaml:12]: same

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/knative/KnativeBroker.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import io.fabric8.kubernetes.api.model.Namespaced;
+import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.model.annotation.Group;
+import io.fabric8.kubernetes.model.annotation.Kind;
+import io.fabric8.kubernetes.model.annotation.Version;
+
+@Version("v1")
+@Group("eventing.knative.dev")
+@Kind("Broker")
+public class KnativeBroker extends CustomResource<KnativeBrokerSpec, KnativeBrokerStatus> implements Namespaced {
```
[the remark this answers is not in the document]
Comment (r00ta) [reply shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/knative/KnativeBroker.java:12]: it does not

File: kustomize/base/shard/resources/deploy.yaml (hunk 5 above)
Comment (r00ta) [reply kustomize/base/shard/resources/deploy.yaml:15]: because it is a workaround for testing. will be removed

File: integration-tests/src/test/java/com/redhat/service/smartevents/integration/tests/steps/IngressSteps.java [hunk 14]
```diff
@@ -47,7 +49,7 @@ public void ingressOfBridgeIsAvailableWithinMinutes(String testBridgeName, int t
                 .pollInterval(Duration.ofSeconds(5))
                 .untilAsserted(() -> IngressResource.optionsJsonEmptyEventResponse(context.getManagerToken(), endpoint)
                         .then()
-                        .statusCode(200));
+                        .statusCode(405));
```
[the remark this answers is not in the document]
Comment (r00ta) [reply integration-tests/src/test/java/com/redhat/service/smartevents/integration/tests/steps/IngressSteps.java:52]: knative does not expose `options`. We just need to check that the ingress is ready so 405 might be ok until we don't find a better solution

File: kustomize/base-openshift/shard/patches/deploy-config.yaml [hunk 15]
```diff
@@ -1,10 +1,16 @@
 apiVersion: v1
 data:
+<<<<<<< MGDOBR-536
```
Comment (jiripetrlik) [inline kustomize/base-openshift/shard/patches/deploy-config.yaml:3]: Is this intentional?

File: kustomize/overlays/minikube/shard/cluster-role.yaml [hunk 16]
```diff
@@ -6,25 +6,26 @@ rules:
...
     verbs:
-      - get
-      - list
+      - "get"
+      - "list"
   - apiGroups:
       - ""
-      - extensions
-      - apps
+      - "extensions"
```
Comment (jiripetrlik) [inline kustomize/overlays/minikube/shard/cluster-role.yaml:21]: Is there any reason to transform all to double qutes?

**Review by jiripetrlik — APPROVED** [review]
Looks good to me. I have only very minor comments. @r00ta Thank you for going through the code on today's meeting.

File: processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/webhook/WebhookActionInvoker.java [hunk 17]
```diff
@@ -49,10 +51,12 @@ private WebhookActionInvoker(String endpoint, WebClient webClient, OidcClient oi
     @Override
     public void onEvent(String event, Map<String, String> headers) {
         HttpRequest<Buffer> request = webClient.postAbs(endpoint);
+        // If the oidcClient is set, then the target is a bridge ingress. The content type has to be application/cloudevents+json.
```
Comment (danielezonca) [inline processor/runtime/src/main/java/com/redhat/service/smartevents/processor/actions/webhook/WebhookActionInvoker.java:54]: We should review this with another ticket in the future

**Review by danielezonca — APPROVED** [review]
Minor comments but LGTM, well done! 👍

File: shard-operator/src/main/resources/templates/bridge-ingress-broker.yaml [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    app.kubernetes.io/created-by: bridge-fleet-shard-operator # Specify the operator
+    app.kubernetes.io/component: ingress # used to specify the component
+  ownerReferences: # used for finding which Processor does this Deployment belong to
+    - apiVersion: ""
+      kind: ""
+      name: ""
+      uid: ""
+  annotations:
+    eventing.knative.dev/broker.class: Kafka
+    x-kafka.eventing.knative.dev/external.topic: ""
```
Comment (matzew) [inline shard-operator/src/main/resources/templates/bridge-ingress-broker.yaml:15]: With 1.23 (release next week) it will be `kafka.eventing.knative.dev/external.topic` (the `x-` was removed).

We do in 1.23 support the `x-` value as well, but I'd recommend changing that asap.

Options:
* wait to merge and change until 1.23 (and matching knative 1.2 midstream) is released
* merge now, change shortly after the release

**Comment by sutaakar** [comment]
a nitpick - this template seems unused any more - https://github.com/r00ta/sandbox/blob/MGDOBR-536/shard-operator/src/main/resources/templates/bridge-ingress-deployment.yaml , to be deleted?

**Comment by r00ta** [comment]
> a nitpick - this template seems unused any more - https://github.com/r00ta/sandbox/blob/MGDOBR-536/shard-operator/src/main/resources/templates/bridge-ingress-deployment.yaml , to be deleted?

Yup, thanks!

**Review by jiripetrlik — APPROVED** [review]
